Fix one of performance regressions in method calling

Watson1978 opened the pull request:
Seems that LIKELY() introduces some penalties at branching.

* Ruby 2.4.1
```
$ ruby -v ~/tmp/bench.rb
ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-darwin16]
       user     system      total        real
   1.460000   0.010000   1.470000 (  1.462540)
```

* Ruby trunk without patch
```
$ ./miniruby -v -Ilib ~/tmp/bench.rb
ruby 2.5.0dev (2017-03-28 trunk 58156) [x86_64-darwin16]
       user     system      total        real
   1.580000   0.000000   1.580000 (  1.579428)
```

* Ruby trunk with patch
```
$ ./miniruby -v -Ilib ~/tmp/bench.rb
ruby 2.5.0dev (2017-03-28 trunk 58156) [x86_64-darwin16]
       user     system      total        real
   1.510000   0.000000   1.510000 (  1.505948)
```

* Test code
```
require 'benchmark'

Benchmark.bm do |x|
  ary = [1,2,3]

  x.report do
    20000000.times do
      ary.at(2)
    end
  end

end
```

Watson1978 committed:
Fix one of performance regressions in method calling

Seems that LIKELY() introduces some penalties at branching.

* Ruby 2.4.1
$ ruby -v ~/tmp/bench.rb
ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-darwin16]
       user     system      total        real
   1.460000   0.010000   1.470000 (  1.462540)

* Ruby trunk without patch
$ ./miniruby -v -Ilib ~/tmp/bench.rb
ruby 2.5.0dev (2017-03-28 trunk 58156) [x86_64-darwin16]
       user     system      total        real
   1.580000   0.000000   1.580000 (  1.579428)

* Ruby trunk with patch
$ ./miniruby -v -Ilib ~/tmp/bench.rb
ruby 2.5.0dev (2017-03-28 trunk 58156) [x86_64-darwin16]
       user     system      total        real
   1.510000   0.000000   1.510000 (  1.505948)

* Test code
require 'benchmark'

Benchmark.bm do |x|
  ary = [1,2,3]

  x.report do
    20000000.times do
      ary.at(2)
    end
  end

end

Watson1978 commented:
Looks like that a lot of `nop` instructions also are invoked around `ruby_th_dtrace_setup()` with clang's generated binary when run same test codes.
![](https://d3uepj124s5rcx.cloudfront.net/items/17023U0q290W1G0F1V1B/170331-0001.png?v=38690013)

Watson1978 committed:
Fix performance regressions in method calling

Seems that UNLIKELY() introduces some penalties at branching.

* Ruby 2.4.1
$ ruby -v ~/tmp/bench.rb
ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-darwin16]
       user     system      total        real
   1.460000   0.010000   1.470000 (  1.462540)

* Ruby trunk with patches
$ ./miniruby -v -Ilib ~/tmp/bench.rb
ruby 2.5.0dev (2017-03-28 trunk 58156) [x86_64-darwin16]
       user     system      total        real
   1.480000   0.000000   1.480000 (  1.479660)

* Test code
require 'benchmark'

Benchmark.bm do |x|
  ary = [1,2,3]

  x.report do
    20000000.times do
      ary.at(2)
    end
  end

end

Watson1978 commented:
Every time, Ruby checks whether DTrace's prove is enabled when it would call a method.
However, DTrace's prove would not provided with dynamically, it is detected to provide at only compilation time.

![](https://d3uepj124s5rcx.cloudfront.net/items/303c370C090f2v0t2V1V/170404-0001.png?v=ba555267)

knu commented:
Just a minor nit, but I think you mean probe, not prove.

Watson1978 commented:
@knu You're right. It was just typo. Thank you.

nurse commented:
This cannot disable probe once it is enabled.

Watson1978 commented:
Looks for me that DTrace's probe will be enabled/disabled with `configure` script.

1. If the system has `dtrace` command, `configure` will mark DTrace is availed.
2. If DTrace is availed, DTrace's probe will be generated with `tool/gen_dummy_probes.rb` at `make` process.

And there is no way to enable/disable DTrace's probe at runtime.

Hmm, I wonder If I missed something.

nurse commented:
There were some detailed documents about this mechanism before, but I couldn't find good document now. Therefore following are abstract of dtrace probes mechanism...

* Whether probes are inserted into the binary or not is statically decided.
  * It depends on the existence of `dtrace` command and configure as you wrote.
  * Those probe is disabled by default (exist in the binary but don't fire event; many `NOP`s you saw are disabled probes) 
* Whether probes fire the event is dynamically decided
  * Users can enable those probes by dtrace command. It dynamically replaces `NOP`s in the memory with the related probe instructions.

Watson1978 commented:
Thank you for replying.
It is very interesting mechanism and it sounds my changing is too bad.

nurse commented:
This still gain performance improvement?
On my environment it doesn't show significant change.

ko1 commented:
It contains two modifications.
Could you separate them?

Method calling is common operation so that we need to care about them.
Please make your benchmarks more carefully.

Watson1978 commented:
```ruby
require 'benchmark/ips'

Benchmark.ips do |x|

  x.report "while" do |iter|
    ary = [1, 2, 3, 4, 5]
    i = 0
    while i < iter do
      ary.at(2)
      i += 1
    end
  end

  x.report "Integer#times" do |iter|
    ary = [1, 2, 3, 4, 5]
    iter.times do
      ary.at(2)
    end
  end

end
```

Watson1978 commented:
## Benchmark result
Using Apple LLVM version 8.1.0 (clang-802.0.41)

### Ruby 2.4.1
```
               while     20.988M (± 2.5%) i/s -    104.955M in   5.004269s
       Integer#times     14.062M (± 3.0%) i/s -     70.306M in   5.004882s
```

### Before (r58192)
```
               while     20.937M (± 2.5%) i/s -    104.668M in   5.002974s
       Integer#times     13.040M (± 1.9%) i/s -     65.238M in   5.005069s
```

* Result : ruby 2.4.1 vs r58192
```
while         -> 0.27 % down
Integer#times -> 7.76 % down
```

### After (r58192 + patches)
```
               while     21.409M (± 3.0%) i/s -    107.055M in   5.005548s
       Integer#times     13.159M (± 2.6%) i/s -     65.822M in   5.005854s
```

* Result : r58192 vs r58192 + patchs
```
while         -> 2.28 % up
Integer#times -> 0.90 % up
```

Watson1978 commented:
## Benchmark result
Using gcc-7 (Homebrew GCC 7.1.0) 7.1.0

### Ruby 2.4.1
```
               while     22.672M (± 2.3%) i/s -    113.292M in   5.000184s
             16.818M (± 1.9%) i/s -     84.169M in   5.006644s
```

### Before (r58192)
```
               while     22.439M (± 2.0%) i/s -    112.208M in   5.002965s
       Integer#times     16.613M (± 3.2%) i/s -     83.029M in   5.003500s
```

* Result : ruby 2.4.1 vs r58192
```
while         -> 0.97 % down
Integer#times -> 1.37 % down
```


### After (r58192 + patches)
```
               while     22.893M (± 1.7%) i/s -    114.492M in   5.002894s
       Integer#times     16.853M (± 1.8%) i/s -     84.302M in   5.004057s
```

* Result : r58192 vs r58192 + patchs
```
while         -> 2.04 % up
Integer#times -> 1.53 % up
```

Watson1978 commented:
Updated benchmark script and result in above.

I found,

1. the regression is occurred in clang + Integer#times combination.
2. The patches in this pull request increase performance very slightly.

So, I guess we need to focus on yield performance such as https://github.com/ruby/ruby/pull/1535

nurse commented:
This improvement is small but this simplifies code.
(whether a patch should be merged or not depends the trade off between improvement and code complexity)
If this is also true on Linux+gcc, I'm positive to merge.
How do you think @ko1?

ko1 commented:
(1) Please explain your environment. You only show your compiler, but you need to show your CPU arch (and OS if it is matter). On this case, "LIKELY" affects branch prediction and CPU arch is a matter (of course, compiler version are also matter). dtrace (systemtap) enable/disable is also matter.
(2) Please show results in a table form. It is easy to compare.
(3) (advanced) Please write the reason why these fix are affect your benchmark results. I'm not sure why such performance changing. I need to understand "why" to approve this PR.
(4) Please explain why you change the benchmark. Benchmark should have the purpose and please make clear what is measured. Additionally, on this case, loop overhead is more matter so you need to measure only method call overhead. This is what I wrote " Please make your benchmarks more carefully." in above message.

Thanks,
Koichi

ko1 commented:
> I need to understand "why" to approve this PR.

In other words, I'll check the reason sometime if nobody explains.

ko1 commented:
I forgot one point.

(5) (related to (3)) It contains two modification. Please separate and show which is matter for performance. I'm not sure about dtrace, but method caching seems good point to use `LIKELY`.

Watson1978 commented:
### 1. Environment
* Hardware : MacBook Pro (Retina, 15-inch, Late 2013)
* CPU : 2.6 GHz Intel Core i7 (Haswell)
* OS : macOS 10.12.5
* dtrace : macOS bundled `dtrace` command by default.
* compiler
    * clang 802.0.4 (in Xcode 8.3.1, maybe it is same with LLVM 3.9.1)
    * gcc 7.1.0
* Other : https://github.com/rugarciap/Turbo-Boost-Switcher to disable Turbo Boost feature in CPU

### 2. Result
* With clang

|  | Ruby 2.4.1 | r58192 | Ruby 2.4.1 vs r58192 |  | Ruby 2.4.1 | r58192 + patches | Ruby 2.4.1 vs r58192 + patches |
| --- | --- | --- | --- | --- | --- | --- | --- |
| while | 104.955M | 104.668M | 0.27% down |  | 104.955M | 107.055M | 2.00% up |
| Integer#times | 70.306M | 65.238M | 7.76% down |  | 70.306M | 65.822M | 6.81% down |

* With gcc

|  | Ruby 2.4.1 | r58192 | Ruby 2.4.1 vs r58192 |  | Ruby 2.4.1 | r58192 + patches | Ruby 2.4.1 vs r58192 + patches |
| --- | --- | --- | --- | --- | --- | --- | --- |
| while | 113.292M | 112.208M | 0.97% down |  | 113.292M | 114.492M | 1.06% up |
| Integer#times | 84.169M | 83.029M | 1.37% down |  | 84.169M | 84.302M | 0.2% up |

### 3. Explain the patch
I've used `Instruments` which is developer tool contained in Xcode to measure the performace for each improvements.

I was guessing that there is regression in method dispatching when I created the patches, and I investigated there.
`Instruments` indicated `vm_exec -> vm_exec_core -> vm_call_cfunc -> Array#at` as function call graph for method dispathing.

![](https://d2ppvlu71ri8gs.cloudfront.net/items/0H3K1e1n2K0o0h190r1Y/function_call_graph.png)

Especially, `vm_exec_core()` spent 4244 msec and `vm_call_cfunc()` spent 1179 msec in 6412 msec execution time and
I thought they have the hotspots for method dispatching.

Next, I investigate `vm_exec_core()` and `vm_call_cfunc()`. In `vm_exec_core()`, the largest hotspot is calling `vm_call_cfunc()` which spent `29.71%` of invoking `vm_exec_core()` (4244 msec). The second hotspot is around `UNLIKELY` which I changed in second patch.
![](https://d2ppvlu71ri8gs.cloudfront.net/items/10082U0b2V1a3b433d0i/170523-0001.png)

If I removed `UNLIKELY`, it remove the second hotspot from `Instruments`.

By same way, I created a first patch for improving `vm_call_cfunc()`.


### 4. Why change the benchmark
The first benchmark script uses Integer#times calling which has huge overhead.
In order to clarify whether really method dispatch has regression, I changed to use `while` loop.

### 5. Separate and show which is matter for performance.
I will try to do later, sorry.

Watson1978 commented:
### 3. Explain the patch (additional)
Under macOS, it has bundled `dtrace` command by default.

I think the condition whether `dtrace` is enabled always returns `true`.
So, I guess `UNLIKELY` inserts some penalty in condition and then I removed.

k0kubun commented:
Could you resolve a conflict? Also we may need to take a benchmark again since things have been changed much.

Watson1978 commented:
@k0kubun done
